sys/net/gnrc/network_layer/ipv6/nib/_nib-internal.c

```diff
@@ -501,28 +504,29 @@ _nib_offl_entry_t *_nib_offl_alloc(const ipv6_addr_t *next_hop, unsigned iface,
         _nib_offl_entry_t *tmp = &_dsts[i];
         _nib_onl_entry_t *tmp_node = tmp->next_hop;
 
-        if ((tmp->pfx_len == pfx_len) &&                /* prefix length matches and */
-            (tmp_node != NULL) &&                       /* there is a next hop that */
-            (_nib_onl_get_if(tmp_node) == iface) &&     /* has a matching interface and */
-            _addr_equals(next_hop, tmp_node) &&         /* equal address to next_hop, also */
-            (ipv6_addr_match_prefix(&tmp->pfx, pfx) >= pfx_len)) {  /* the prefix matches */
-            /* exact match (or next hop address was previously unset) */
-            DEBUG("  %p is an exact match\n", (void *)tmp);
-            if (next_hop != NULL) {
-                memcpy(&tmp_node->ipv6, next_hop, sizeof(tmp_node->ipv6));
+        if (tmp->mode != _EMPTY) {
```

Comment:
would that work?
```suggestion
        if (tmp->mode == _EMPTY) {
            dst = tmp;
            continue;
        }
```
But tbh I don't know what's the difference between `tmp->next_hop == NULL` and `tmp->mode == _EMPTY`, I would have thought `tmp->next_hop == NULL` could also happen on a non-empty entry.

Reply:
Even for an on-link prefix, `tmp->next_hop` points to an NCE in order to refer to an interface, so I wouldn't know a case where `tmp->next_hop == NULL` while `tmp->mode != _EMPTY`.

Reply:
I think your suggested change would use the last free entry instead of the first one, when there is no exact matching entry. I think, while it wouldn't cause errors, it's an unusual allocation policy (compared to all other allocations in GNRC).

Reply:
```C
        if (tmp->mode == _EMPTY) {
            if (dst == NULL) {
                dst = tmp;
            }
            continue;
        }
```

would also work, at the expense of one more branch that might be a bit unececary.
What I'm getting at was reducing the level of indentation if you return / continue early.

Reply:
I think every offlink entry has an onlink entry in our implementation.
Even if it only stores an interface ID and has an unspecified next hop..
In `_nib_offl_alloc`, if `dst` was `NULL` this means no offlink entry could be allocated and none matched.

Reply:
To be safe, an `assert(tmp_node)` when the prefix matches does not harm.

Reply:
> if dst was NULL this means no offlink entry could be allocated and none matched

That is true only after all entries were iterated through. The change suggested by @benpicco is however in the loop still, in which case `dst == NULL` means no candidate has been found _so far_ (in previous iterations).